### DISCUSSION TITLE
download_model() as option for __init__...

### DISCUSSION THREAD

**Discussion by IIMunchII** [alexandrainst/danlp#119] [Ideas]
When initialising models we sometimes don't have access to the web. Having the download_model funktion inside a class `__init__()` creates a big firewall dependency for all people working on premise or behind a firewall. 

Being able to point directly to a local folder or any on prem repository where the model objects are hosted, is essential for the usage of this library in a generel production setting.

I propose a change in all model classes `__init__()` methods, that separates the generel `self.model_path` (or several paths) attribute from the `download_model()` function.
 
Right now the `BertBase` init method looks like this.

```python
    def __init__(self, cache_dir=DEFAULT_CACHE_DIR, verbose=False):
        from transformers import BertTokenizer, BertModel
        import torch
        # download model
        self.path_model= download_model('bert.botxo.pytorch', cache_dir, process_func=_unzip_process_func,verbose=verbose)
        # Load pre-trained model tokenizer
        self.tokenizer = BertTokenizer.from_pretrained(self.path_model)
        # Load pre-trained model (weights)
        self.model = BertModel.from_pretrained(self.path_model,
                                          output_hidden_states = True, # Whether the model returns all hidden-states.
                                          )
```

We could instead do something like this.

```python
    def __init__(self, 
                 cache_dir=DEFAULT_CACHE_DIR, 
                 download_model=True,
                 verbose=False,
                 **kwargs):
        from transformers import BertTokenizer, BertModel
        import torch
        if download_model:
            self.path_model = download_model('bert.botxo.pytorch',
                                             cache_dir,
                                             process_func=_unzip_process_func, 
                                             verbose=verbose)
        else:
            self.path_model = kwargs.get("path_model")
        # Load pre-trained model tokenizer
        self.tokenizer = BertTokenizer.from_pretrained(self.path_model)
        # Load pre-trained model (weights)
        self.model = BertModel.from_pretrained(self.path_model,
                                               output_hidden_states = True, # Whether the model returns all hidden-states.
                                              )
```

Or maybe more comprehensive change that changes the API to init the model class without downloading or setting path, and then call the download method on the instance as a way for the user to setup/load the model before usage.
Another idea would be to write a class called ModelStorage and use that class to specify where the models are located or whether they should be downloaded before loading the model classes.

**Comment by ophelielacroix** [comment]
Hi Jonas,
thanks for your input and suggestion. It is indeed very relevant, so we added it to our roadmap. 
This type of feedback is really valuable to us in order to improve the package!
